src/pool-bin/libraries/BinPool.sol

```diff
@@ -184,51 +191,64 @@ library BinPool {
         uint16 binStep;
     }
 
+    struct SwapState {
+        uint24 activeId;
+        uint24 protocolFee;
+        uint24 swapFee;
+        bytes32 feeForProtocol;
+    }
+
     function swap(State storage self, SwapParams memory params, uint128 amountIn)
         internal
-        returns (BalanceDelta result, bytes32 feeForProtocol, uint24 activeId, uint24 swapFee)
+        returns (BalanceDelta result, SwapState memory swapState)
     {
-        if (amountIn == 0) revert BinPool__InsufficientAmountIn();
```

Comment:
can consider moving this check up `if (amountIn == 0) revert BinPool__InsufficientAmountIn();` so in revert case it will consume less gas